### PR TITLE
*: Improve test for push down.

### PR DESCRIPTION
**Pull request by hanfei1991**
Make the test about *** push-down more readable.
PTAL @shenli @coocood @zimulala @XuHuaiYu

### CODE REVIEW

File: plan/expr_to_pb.go [hunk 1]
```diff
@@ -24,13 +24,14 @@ import (
 	"github.com/pingcap/tipb/go-tipb"
 )
 
-func expressionsToPB(exprs []expression.Expression, client kv.Client) (pbExpr *tipb.Expr, remained []expression.Expression) {
+func expressionsToPB(exprs []expression.Expression, client kv.Client) (pbExpr *tipb.Expr, remained []expression.Expression, pushed []expression.Expression) {
```
Comment (shenli) [inline plan/expr_to_pb.go]: Why add new return value?

Comment (hanfei1991) [inline plan/expr_to_pb.go]: We need to know the expressions that be pushed down for tests.

File: plan/plan_test.go [hunk 2]
```diff
@@ -317,17 +387,22 @@ func (s *testPlanSuite) TestOperatorsPushDown(c *C) {
...
 		{
 			sql:    "a and b",
-			cond:   "and(test.t.a, test.t.b)",
+			cond:   "test.t.b",
 			exprPB: "\b\xc9\x01\x12\b\x80\x00\x00\x00\x00\x00\x00\x02",
 		},
 		{
+			sql:    "a or (b and c)",
+			cond:   "or(test.t.a, and(test.t.b, test.t.c))",
+			exprPB: "\b\xfe\x11\x1a\r\b\xc9\x01\x12\b\x80\x00\x00\x00\x00\x00\x00\x01\x1a!\b\xfd\x11\x1a\r\b\xc9\x01\x12\b\x80\x00\x00\x00\x00\x00\x00\x02\x1a\r\b\xc9\x01\x12\b\x80\x00\x00\x00\x00\x00\x00\x03",
```
Comment (coocood) [inline plan/plan_test.go]: We don't need this to be showed.

File: plan/expr_to_pb.go (hunk 1 above)
Comment (coocood) [inline plan/expr_to_pb.go]: The first return value `pbExpr` is also pushed, I think move `pushed` before `remained` is better.

File: plan/plan_test.go [hunk 3]
```diff
@@ -385,53 +351,86 @@ func (s *testPlanSuite) TestOperatorsPushDown(c *C) {
...
+				break
 			}
 			p = p.GetChildByIndex(0)
 		}
 	}
 }
 
-// TestBuiltinFuncsPushDown tests whether logic operators been pushed down successfully.
-func (s *testPlanSuite) TestBuiltinFuncsPushDown(c *C) {
+// TestPushDownExpression tests whether logic operators been pushed down successfully.
```
Comment (XuHuaiyu) [inline plan/plan_test.go]: update comment, it is not logic operators.

File: plan/physical_plans.go [hunk 4]
```diff
@@ -117,8 +145,14 @@ func (p *physicalTableSource) addTopN(prop *requiredProperty) bool {
 	}
 	count := int64(prop.limit.Count + prop.limit.Offset)
 	p.LimitCount = &count
-	for _, item := range prop.props {
-		p.SortItems = append(p.SortItems, sortByItemToPB(p.client, item.col, item.desc))
+	for _, prop := range prop.props {
+		item := sortByItemToPB(p.client, prop.col, prop.desc)
+		if item == nil {
+			p.clearForTopnPushDown()
+			return false
```
Comment (XuHuaiyu) [inline plan/physical_plans.go]: why not define the return value of this function?

File: plan/physical_plans.go [hunk 5]
```diff
@@ -70,22 +67,41 @@ type physicalTableSource struct {
...
+
+	// The following fields are only for tests.
+	aggFuncs   []expression.AggregationFunction
+	gbyItems   []expression.Expression
+	sortItems  []*ByItems
+	conditions []expression.Expression
 }
 
-func (p *physicalTableSource) clear() {
+func (p *physicalTableSource) clearForAggPushDown() {
```
Comment (XuHuaiyu) [inline plan/physical_plans.go]: comment for when this function works.

File: plan/physical_plans.go [hunk 6]
```diff
@@ -70,22 +67,41 @@ type physicalTableSource struct {
...
-	p.GbyItems = nil
+	p.AggFuncsPB = nil
+	p.GbyItemsPB = nil
 	p.Aggregated = false
+
+	p.aggFuncs = nil
+	p.gbyItems = nil
+}
+
+func (p *physicalTableSource) clearForTopnPushDown() {
```
Comment (XuHuaiyu) [inline plan/physical_plans.go]: ditto

File: plan/physical_plans.go (hunk 4 above)
Comment (hanfei1991) [inline plan/physical_plans.go]: I didn't catch your point. You mean define the return value of clear***? It makes no sense.

**Comment by hanfei1991** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
LGTM

File: plan/physical_plans.go [hunk 7]
```diff
@@ -97,6 +113,18 @@ func needValue(af expression.AggregationFunction) bool {
 		af.GetName() == ast.AggFuncMax || af.GetName() == ast.AggFuncMin || af.GetName() == ast.AggFuncGroupConcat
 }
 
+func (p *physicalTableSource) tryToAddUnionScan(resultPlan PhysicalPlan) PhysicalPlan {
+	if !p.readOnly {
```
Comment (shenli) [inline plan/physical_plans.go]: if p.readOnly {
return 
}

File: plan/physical_plans.go [hunk 8]
```diff
@@ -70,22 +67,41 @@ type physicalTableSource struct {
...
 
-	LimitCount *int64
-	SortItems  []*tipb.ByItem
+	// AccessCondition is used to calculate range.
+	AccessCondition []expression.Expression
+
+	LimitCount  *int64
+	SortItemsPB []*tipb.ByItem
+
+	// The following fields are only for tests.
```
Comment (shenli) [inline plan/physical_plans.go]: Update this comment.

**Comment by shenli** [comment]
LGTM

**Pull request merged**